packages/container/lib/container.js

```diff
@@ -466,3 +575,56 @@ function buildFakeContainerFunction(container, containerProperty, ownerProperty)
     return container[containerProperty](...arguments);
   };
 }
+
+class DeprecatedFactoryManager {
+  constructor(container, factory, fullName) {
+    this.container = container;
+    this.class = factory;
+    this.fullName = fullName;
+  }
+
+  create(props = {}) {
+    return instantiate(this.class, props, this.container, this.fullName);
+  }
+}
+
+class FactoryManager {
+  constructor(container, factory, fullName, normalizedName) {
+    this.container = container;
+    this.class = factory;
+    this.fullName = fullName;
+    this.normalizedName = normalizedName;
+  }
+
+  create(options = {}) {
+    let injections = injectionsFor(this.container, this.normalizedName);
+    let props = assign({}, injections, options);
+
+    runInDebug(() => {
+      let lazyInjections;
+      let validationCache = this.container.validationCache;
+      // Ensure that all lazy injections are valid at instantiation time
+      if (!validationCache[this.fullName] && this.class && typeof this.class._lazyInjections === 'function') {
+        lazyInjections = this.class._lazyInjections();
+        lazyInjections = this.container.registry.normalizeInjectionsHash(lazyInjections);
+
+        this.container.registry.validateInjections(lazyInjections);
+      }
+
+      validationCache[this.fullName] = true;
+    });
+
+    this.class[NAME_KEY] = this.container.registry.makeToString(this.class, this.fullName);
```

Comment:
It does looks like this should be added to the injections [per master](https://github.com/emberjs/ember.js/blob/e601285bed0e2e108d41a68a61dc2098aac26dd7/packages/container/lib/container.js#L273) instead of set on the class.